pydeseq2/dds.py

```diff
@@ -213,55 +226,153 @@ def __init__(
-        # Stored in the obsm attribute of the dataset
-        self.obsm["design_matrix"] = build_design_matrix(
-            metadata=self.obs,
-            design_factors=self.design_factors,
-            continuous_factors=self.continuous_factors,
-            ref_level=ref_level,
-            expanded=False,
-            intercept=True,
+        self.single_design_factors = list(
+            set(self.design_factors) & set(self.obs.columns)
         )
+        # Remark it works if self.design_factors is a string as well by iterating
+        # on chars
+        self.design_factors = replace_underscores(self.design_factors)
+
+        self.continuous_factors = continuous_factors
+
+        if self.obs[self.single_design_factors].isna().any().any():
+            raise ValueError("NaNs are not allowed in the design factors.")
+        self.obs[self.single_design_factors] = self.obs[
+            self.single_design_factors
+        ].astype(str)
+
+        # Build the design matrix or use given one
+        # Stored in the obsm attribute of the dataset
+        if design_matrix is None:
+            self.obsm["design_matrix"] = build_design_matrix(
+                metadata=self.obs,
+                design_factors=self.design_factors,
+                continuous_factors=self.continuous_factors,
+                ref_level=ref_level,
+                expanded=False,
+                intercept=True,
+            )
+        else:
+            warnings.warn(
+                """Design matrix was given; ignoring design_factors and
+                ref_level""",
+                UserWarning,
+                stacklevel=2,
+            )
+            if len(design_matrix.index) != len(self.obs.index):
+                raise ValueError("Design matrix and metadata have different lengths")
+            if "intercept" not in design_matrix.columns:
+                raise ValueError("Design matrix doesn't have intercept")
+
+            if not design_matrix["intercept"].equals(
+                pd.Series(1.0, index=design_matrix.index)
+            ):
+                raise ValueError("Design matrix doesn't have intercept")
+            for individual_factor_or_level in itertools.chain(
+                *[
+                    re.sub(r"\:|_vs_|_", " ", col).split(" ")
+                    for col in design_matrix
+                    if col != "intercept"
+                ]
+            ):
+                if (
+                    individual_factor_or_level not in self.obs.columns
+                    and individual_factor_or_level
+                    not in itertools.chain(
+                        *[self.obs[col].unique().tolist() for col in self.obs.columns]
+                    )
+                ):
+                    raise ValueError(
+                        "Design matrix contains unknown factor"
+                        f" {individual_factor_or_level}"
+                    )
+
+            # With some luck here design_matrix is valid all the time
```

Comment:
What do you mean?

Reply:
I tried to add conditions but feel free to comment